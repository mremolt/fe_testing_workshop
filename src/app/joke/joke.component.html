<h1>Random joke for category "{{ category }}"</h1>

<div *ngIf="joke$ | async as joke; else noJoke">
  <p class="joke">{{ joke.value }}</p>

  <button class="reload-button" mat-raised-button (click)="reload()">load new random joke</button>
</div>

<ng-template #noJoke>
  <p>No joke loaded.</p>
</ng-template>

<p>
  <a [routerLink]="['/jokes']">back</a>
</p>
