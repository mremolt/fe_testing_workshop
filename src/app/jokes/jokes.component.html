<h1>Joke Categories</h1>

<p>Select a category:</p>

<mat-nav-list class="categories">
  <a
    mat-list-item
    class="category"
    [routerLink]="['/', 'jokes', category]"
    *ngFor="let category of categories$ | async"
  >
    {{ category }}
  </a>
</mat-nav-list>
