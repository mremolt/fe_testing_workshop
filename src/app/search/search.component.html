<h1>Search for Jokes</h1>

<form class="example-form" [formGroup]="form" (submit)="search(form.value.query)" name="form">
  <mat-form-field class="example-full-width" appearance="fill">
    <mat-label>Search for text content</mat-label>
    <input type="search" matInput placeholder="e.g. pain, kick" formControlName="query" />
  </mat-form-field>

  <button [disabled]="form.invalid" mat-raised-button color="primary">Search</button>
</form>

<mat-list>
  <mat-list-item *ngFor="let joke of jokes$ | async">
    <p class="joke" matLine>{{ joke.value }}</p>
    <p matLine>
      Categories: <span *ngFor="let category of joke.categories">{{ category }}</span>
    </p>
  </mat-list-item>
</mat-list>
