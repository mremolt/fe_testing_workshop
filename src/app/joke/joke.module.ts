import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { MatLegacyButtonModule as MatButtonModule } from '@angular/material/legacy-button';

import { JokeComponent } from './joke.component';

@NgModule({
  declarations: [JokeComponent],
  imports: [CommonModule, RouterModule, MatButtonModule],
})
export class JokeModule {}
