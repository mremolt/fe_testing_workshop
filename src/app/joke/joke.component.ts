import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Observable } from 'rxjs';
import { Joke } from '../models/joke.interface';
import { JokesService } from '../services/jokes.service';

@Component({
  selector: 'app-joke',
  templateUrl: './joke.component.html',
  styleUrls: ['./joke.component.scss'],
})
export class JokeComponent implements OnInit {
  public joke$?: Observable<Joke>;
  public category?: string;

  constructor(private readonly route: ActivatedRoute, private readonly jokesService: JokesService) {}

  public ngOnInit(): void {
    this.route.params.subscribe((params) => {
      this.category = params['category'];

      if (this.category) {
        this.loadJoke();
      }
    });
  }

  public reload(): void {
    this.loadJoke();
  }

  private loadJoke(): void {
    this.joke$ = this.jokesService.getRandomJoke(this.category);
  }
}
