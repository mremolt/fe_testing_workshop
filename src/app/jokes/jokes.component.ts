import { Component, OnInit } from '@angular/core';
import { AbstractControl, FormBuilder, FormGroup, ValidatorFn } from '@angular/forms';
import { Observable } from 'rxjs';
import { JokesService } from '../services/jokes.service';

export const validateCategory: ValidatorFn = (control: AbstractControl) => {
  const value = String(control.value);

  if (value.length < 3) {
    return { tooSmall: true, actual: value.length, required: 3 };
  }

  if (value.length > 8) {
    return { tooLarge: true, actual: value.length, required: 8 };
  }

  return null;
};

@Component({
  selector: 'app-jokes',
  templateUrl: './jokes.component.html',
  styleUrls: ['./jokes.component.scss'],
})
export class JokesComponent implements OnInit {
  public categories$: Observable<Array<string>>;
  public readonly form: FormGroup;

  constructor(private readonly jokesService: JokesService, private readonly fb: FormBuilder) {
    this.form = this.fb.group({
      category: ['', [validateCategory]],
    });

    this.categories$ = this.jokesService.getCategories();
  }

  public ngOnInit(): void {}
}
