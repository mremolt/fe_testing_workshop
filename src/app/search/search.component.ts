import { Component } from '@angular/core';
import { UntypedFormBuilder, UntypedFormGroup } from '@angular/forms';
import { Observable } from 'rxjs';

import { Joke } from '../models/joke.interface';
import { JokesService } from '../services/jokes.service';
import { validateSearch } from '../utils/validators';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss'],
})
export class SearchComponent {
  public jokes$?: Observable<Array<Joke>>;
  public form = this.fb.group({
    query: ['', validateSearch],
  });

  constructor(private readonly jokesService: JokesService, private readonly fb: UntypedFormBuilder) {}

  public search(query: string): void {
    this.jokes$ = this.jokesService.search(query);
  }
}
