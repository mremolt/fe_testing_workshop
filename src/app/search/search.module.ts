import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { MatLegacyFormFieldModule as MatFormFieldModule } from '@angular/material/legacy-form-field';
import { MatLegacyInputModule as MatInputModule } from '@angular/material/legacy-input';
import { MatLegacyButtonModule as MatButtonModule } from '@angular/material/legacy-button';
import { MatLegacyListModule as MatListModule } from '@angular/material/legacy-list';

import { SearchComponent } from './search.component';

@NgModule({
  declarations: [SearchComponent],
  imports: [
    CommonModule,
    RouterModule,
    ReactiveFormsModule,
    MatFormFieldModule,
    MatInputModule,
    MatButtonModule,
    MatListModule,
  ],
})
export class SearchModule {}
