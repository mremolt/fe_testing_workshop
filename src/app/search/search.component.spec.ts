import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { fireEvent, render, RenderResult, screen } from '@testing-library/angular';
import userEvent from '@testing-library/user-event';

import { mock } from 'jest-mock-extended';
import { lastValueFrom, of } from 'rxjs';
import { Joke } from '../models/joke.interface';
import { JokesService } from '../services/jokes.service';

import { SearchComponent } from './search.component';

describe('SearchComponent', () => {
  const jokesServiceMock = mock<JokesService>();

  const jokesMock: Array<Joke> = [
    {
      categories: [],
      created_at: '2020-01-05 13:42:30.480041',
      icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
      id: 'Rjq1W2u8S3uYKDMqfIqVlQ',
      updated_at: '2020-01-05 13:42:30.480041',
      url: 'https://api.chucknorris.io/jokes/Rjq1W2u8S3uYKDMqfIqVlQ',
      value: 'Chuck Norris once nearly drowned an ocean in a fit of rage.',
    },
    {
      categories: [],
      created_at: '2020-01-05 13:42:28.143137',
      icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
      id: 'wYElYsATTVWzI-3__zGwPw',
      updated_at: '2020-01-05 13:42:28.143137',
      url: 'https://api.chucknorris.io/jokes/wYElYsATTVWzI-3__zGwPw',
      value: 'Chuck Norris died 20 years ago but death just hasnt built up the courage to tell him yet',
    },
  ];

  jokesServiceMock.search.mockReturnValue(of(jokesMock));

  describe('SearchComponent API', () => {
    let subject: SearchComponent;

    beforeEach(() => {
      subject = new SearchComponent(jokesServiceMock, new FormBuilder());
    });

    it('should create', () => {
      expect(subject).toBeTruthy();
    });

    describe('form', () => {
      it('should publish form property', () => {
        expect(subject.form.value).toMatchSnapshot();
      });

      it('should validate the form', () => {
        expect(subject.form.valid).toBeFalsy();

        subject.form.setValue({ query: 'secret' });

        expect(subject.form.valid).toBeTruthy();
      });
    });

    describe('search()', () => {
      it('should fill the jokes$ property', async () => {
        subject.search('rage');

        const result = await lastValueFrom(subject.jokes$!);

        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('SearchComponent', () => {
    let subject: RenderResult<SearchComponent>;
    let component: SearchComponent;

    beforeEach(async () => {
      subject = await render(SearchComponent, {
        imports: [ReactiveFormsModule],
        providers: [{ provide: JokesService, useFactory: () => jokesServiceMock }],
        schemas: [CUSTOM_ELEMENTS_SCHEMA],
      });
      component = subject.fixture.componentInstance;
    });

    it('should render', () => {
      expect(screen.getByText('Search for Jokes')).toBeInTheDocument();
    });

    describe('search form', () => {
      it('should connect the form to the template', async () => {
        const searchField = screen.getByPlaceholderText('e.g. pain, kick');

        await userEvent.type(searchField, 'Sylvester');
        await fireEvent.blur(searchField);

        expect(component.form.value).toMatchSnapshot();
      });

      it('should trigger the search on submit', async () => {
        const searchField = screen.getByPlaceholderText('e.g. pain, kick');
        const searchButton = screen.getByText('Search');

        await userEvent.type(searchField, 'kick');
        await fireEvent.blur(searchField);
        await fireEvent.click(searchButton);

        expect(jokesServiceMock.search).toHaveBeenCalledWith('kick');
      });

      it('should not trigger the search on submit for an invalid form', async () => {
        const searchField = screen.getByPlaceholderText('e.g. pain, kick');
        const searchButton = screen.getByText('Search');

        await userEvent.type(searchField, 'this is too long');
        await fireEvent.blur(searchField);
        await fireEvent.click(searchButton);

        expect(jokesServiceMock.search).not.toHaveBeenCalled();
      });
    });
  });
});
