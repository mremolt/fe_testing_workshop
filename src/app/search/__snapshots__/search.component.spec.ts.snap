// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent SearchComponent API form should publish form property 1`] = `
{
  "query": "",
}
`;

exports[`SearchComponent SearchComponent API search() should fill the jokes$ property 1`] = `
[
  {
    "categories": [],
    "created_at": "2020-01-05 13:42:30.480041",
    "icon_url": "https://assets.chucknorris.host/img/avatar/chuck-norris.png",
    "id": "Rjq1W2u8S3uYKDMqfIqVlQ",
    "updated_at": "2020-01-05 13:42:30.480041",
    "url": "https://api.chucknorris.io/jokes/Rjq1W2u8S3uYKDMqfIqVlQ",
    "value": "Chuck Norris once nearly drowned an ocean in a fit of rage.",
  },
  {
    "categories": [],
    "created_at": "2020-01-05 13:42:28.143137",
    "icon_url": "https://assets.chucknorris.host/img/avatar/chuck-norris.png",
    "id": "wYElYsATTVWzI-3__zGwPw",
    "updated_at": "2020-01-05 13:42:28.143137",
    "url": "https://api.chucknorris.io/jokes/wYElYsATTVWzI-3__zGwPw",
    "value": "Chuck Norris died 20 years ago but death just hasnt built up the courage to tell him yet",
  },
]
`;

exports[`SearchComponent SearchComponent search form should connect the form to the template 1`] = `
{
  "query": "Sylvester",
}
`;
